### PR TITLE
Add a new CreateUser function to return CleintKey

### LINKED ISSUE #24: Client ID is generated, but nothing is returned
Client ID is used with the entertainment API. we should expose or return the value.

https://developers.meethue.com/develop/hue-entertainment/philips-hue-entertainment-api/

### PR DESCRIPTION
**Pull request by SteelPhase**
There were some go format changes, and some test updates.
I've also updated CreateUser to not generate a client key, as it doesn't return it, and it's never used.

Fixes #24

### CODE REVIEW

**Comment by SteelPhase** [comment]
Looks like something changed after go 1.14 that broke some unit tests. I fixed them, and they pass for 1.14, and 1.15 now. Any suggestions on how to resolve issues with 1.13

File: .travis.yml [hunk 1]
```diff
@@ -2,9 +2,9 @@ os: linux
 dist: xenial
 language: go
 go:
-  - 1.11.x
-  - 1.12.x
   - 1.13.x
+  - 1.14.x
+  - 1.15.x
```
Comment (amimof) [inline .travis.yml:7]: No need to modify the CI in this PR. Please open another PR to make those changes.

File: bridge.go [hunk 2]
```diff
@@ -17,6 +17,12 @@ type Bridge struct {
 	ID   string `json:"id,omitempty"`
 }
 
+// CreatedUser represents a created user from the Configuration API
+type CreatedUser struct {
+	Username  string `json:"username,omitempty"`
+	ClientKey string `json:"clientkey,omitempty"`
+}
```
Comment (amimof) [inline bridge.go:24]: No need for this struct. We can simply put `ClientKey` as a field to the existing `Whitelist` struct.

File: bridge.go [hunk 3]
```diff
@@ -130,6 +137,60 @@ func (b *Bridge) CreateUserContext(ctx context.Context, n string) (string, error
 
 }
 
+// CreateUser2 creates a user by adding deviceType to the list of whitelisted users on the bridge.
+// The link button on the bridge must have been pressed before calling CreateUser.
+func (b *Bridge) CreateUser2(deviceType string, generateClientKey bool) (*CreatedUser, error) {
```
Comment (amimof) [inline bridge.go:142]: Not a very descriptive function name. I suggest using `CreateUserWithClientKey` to make it specific to the functions implementation. And drop `generateClientKey`.

**Review by amimof — CHANGES_REQUESTED** [review]
Please have a look at my comments

**Comment by amimof** [comment]
Hi @SteelPhase thanks for contributing. I made a couple of comments to your changes, have a look. I believe we can re-use the existing `Whitelist` struct as well as making the addition of `CreateUser` more specific to its purpose.

File: .travis.yml (hunk 1 above)
Comment (SteelPhase) [reply .travis.yml:7]: Yeah I'll take that out. Was just trying to figure out why the tests were passing for me on my machine. Was worried it was a Windows 10 issue

File: bridge.go (hunk 2 above)
Comment (SteelPhase) [reply bridge.go:24]: I'll take a look at that struct, I believe you are correct

File: bridge.go (hunk 3 above)
Comment (SteelPhase) [reply bridge.go:142]: Yeah that sounds good with me. I can make those changes.

**Comment by SteelPhase** [comment]
Unbroke the tests, and fixed the versions of go tested. Updated the name for CreateUser2. Let me know what you think

File: bridge.go [hunk 4]
```diff
@@ -84,22 +84,23 @@ func (b *Bridge) GetConfigContext(ctx context.Context) (*Config, error) {
 
 }
 
-// CreateUser creates a user by adding n to the list of whitelists in the bridge.
+// CreateUser creates a user by adding deviceType to the list of whitelisted users on the bridge.
 // The link button on the bridge must have been pressed before calling CreateUser.
-func (b *Bridge) CreateUser(n string) (string, error) {
```
Comment (amimof) [inline bridge.go:89]: No need to change existing function including signature and comment.

File: bridge.go [hunk 5]
```diff
@@ -84,22 +84,23 @@ func (b *Bridge) GetConfigContext(ctx context.Context) (*Config, error) {
...
-	return b.CreateUserContext(context.Background(), n)
+// Note: A client key is not generated with this method
+func (b *Bridge) CreateUser(deviceType string) (string, error) {
+	return b.CreateUserContext(context.Background(), deviceType)
 }
 
-// CreateUserContext creates a user by adding n to the list of whitelists in the bridge.
+// CreateUserContext creates a user by adding deviceType to the list of whitelisted users on the bridge.
 // The link button on the bridge must have been pressed before calling CreateUser.
-func (b *Bridge) CreateUserContext(ctx context.Context, n string) (string, error) {
```
Comment (amimof) [inline bridge.go:95]: Same here. No need to change existing function including signature and comment.

**Review by amimof — CHANGES_REQUESTED** [review]

**Comment by amimof** [comment]
@SteelPhase 
 Please have a look at my review comments. Something else I want to add is that after your changes, the current `CreateUser` might as well make use of `CreateUserWithClientKey` but just return the Username. Right now they both do the same thing but return different types.

**Comment by SteelPhase** [comment]
I see what your are saying. I'm going to take an alternate approach. let me know what you think

**Comment by amimof** [comment]
Should be quite simple really. Your new functions should basically replace the current CreateUser* functions and. And for example `CreateUserContext` would then do something like:
```Go
func (b *Bridge) CreateUserContext(ctx context.Context, n string) (string, error) {
	w, err := CreateUserWithClientKeyContext(ctx, n)
	if err != nil {
		return err, ""
	}
	return w.ClientKey, nil
}
```
And we can figure out the rest from there

**Comment by SteelPhase** [comment]
I just pushed a new commit. Tried to make it as clean as possible. I am going to test them in a few to see if they work as expected with my hue bridge. Just tied up in a meeting

**Comment by amimof** [comment]
I almost got it. Look at my example. Let your new functions do what `CreateUser` and `CreateUserContext` do today. You've basically done it the other way around. This way we don't break stuff for other users which are using CreateUser functions currently. You can ping me on the Kubernetes slack if you have any further doubts.

**Comment by amimof** [comment]
@SteelPhase consider this
```Go
func (b *Bridge) CreateUser(n string) (string, error) {
	return b.CreateUserContext(context.Background(), n)
}

func (b *Bridge) CreateUserContext(ctx context.Context, n string) (string, error) {
	w, err := b.CreateUserWithClientKeyContext(context.Background(), deviceType)
	if err != nil {
		return nil, err
	}
	return w.ClientKey
}

func (b *Bridge) CreateUserWithClientKeyContext(ctx context.Context) (*Whitelist, error) {
	var a []*APIResponse

	body := struct {
		DeviceType        string `json:"devicetype,omitempty"`
		GenerateClientKey bool   `json:"generateclientkey,omitempty"`
	}{n, true}
	
	...

	return &wl, nil
}

func (b *Bridge) CreateUserWithClientKey(deviceType string) (*Whitelist, error) {
	return b.CreateUserWithClientKeyContext(context.Background(), deviceType)
}


```

**Comment by SteelPhase** [comment]
I don't see how what you've posted as an example is any different than what I've committed. I have the logic in a private function, which is no different from what you have done. Your CreateUserContext would also literally break any person using the CreateUser funciton, as it's returning the wrong this. You should be returin wl.Username...

**Comment by amimof** [comment]
If you don’t  see it then i can’t  help you any further unfortunately. The code I suggested was merely an example of the structure. So of course you might expect typos in the details.

**Comment by SteelPhase** [comment]
you aren't explaining your self, and there's no obvious difference.

**Comment by amimof** [comment]
I’ve explained in every way possible and even given you two different code examples. There are differences in terms of code design and DRY:nes which your contribution lacks.

**Comment by SteelPhase** [comment]
There's nothing repetitive in what I did. I'm sorry you didn't bother to actually look at what I wrote. Feel free to implement your self. Just trying to help out a project, but you've made it perfectly clear you don't want it.

**Pull request closed without merging**